internal/application/application.go

```diff
@@ -494,8 +568,127 @@ func Logs(ctx context.Context, logChan chan tailer.ContainerLogLine, wg *sync.Wa
 	return tailer.FetchLogs(ctx, logChan, wg, config, cluster)
 }
 
-// fetch is a common helper for Lookup and List. It fetches all information about an
-// application from the cluster.
+// aggregate is an internal helper for List. It merges the information from an application resource
+// and adjacent secrets, pods, metrics, etc. into a proper application structure.
+
+func aggregate(ctx context.Context,
+	cluster *kubernetes.Cluster,
+	appCR unstructured.Unstructured,
+	scales map[string]v1.Secret,
+	bounds map[string]v1.Secret,
+	envs map[string]v1.Secret,
+	pods map[string][]v1.Pod,
+	routes map[string][]string,
+	metrics map[string]metricsv1beta1.PodMetrics,
+) (*models.App, error) {
+	key := ConfigurationKey(appCR.GetName(), appCR.GetNamespace())
+
+	appPods := pods[key]
+	appRoutes := routes[key]
+
+	appName := appCR.GetName()
+	namespace := appCR.GetNamespace()
+	meta := models.NewAppRef(appName, namespace)
+	app := meta.App()
+
+	// I. Unpack the auxiliary data in the various secrets
+
+	env, found := envs[key]
+	if !found {
+		return nil, errors.New("finding env")
```

Comment:
Is it an err if no env exists (same for other resources)?

Reply:
Yes, even if no env assignments are made, the secret has to exist.
Ditto for the others.

We only might have a temporary condition during app deletion where not all pieces of an app may exist anymore, before it is completely gone.